src/Microsoft.FeatureManagement/ServiceCollectionExtensions.cs

```diff
@@ -16,23 +19,35 @@ namespace Microsoft.FeatureManagement
     public static class ServiceCollectionExtensions
     {
         /// <summary>
-        /// Adds required feature management services and built-in feature filters.
+        /// Adds required feature management services.
         /// </summary>
         /// <param name="services">The service collection that feature management services are added to.</param>
         /// <returns>A <see cref="IFeatureManagementBuilder"/> that can be used to customize feature management functionality.</returns>
         public static IFeatureManagementBuilder AddFeatureManagement(this IServiceCollection services)
         {
             services.AddLogging();
 
+            services.AddMemoryCache();
+
             //
             // Add required services
             services.TryAddSingleton<IFeatureDefinitionProvider, ConfigurationFeatureDefinitionProvider>();
 
-            services.AddSingleton<IFeatureManager, FeatureManager>();
+            services.AddSingleton(sp => new FeatureManager(
+                sp.GetRequiredService<IFeatureDefinitionProvider>(),
+                sp.GetRequiredService<IOptions<FeatureManagementOptions>>().Value)
+            {
+                FeatureFilters = sp.GetRequiredService<IEnumerable<IFeatureFilterMetadata>>(),
+                SessionManagers = sp.GetRequiredService<IEnumerable<ISessionManager>>(),
+                Cache = sp.GetRequiredService<IMemoryCache>(),
+                Logger = sp.GetRequiredService<ILoggerFactory>().CreateLogger<FeatureManager>()
+            });
+
+            services.TryAddSingleton<IFeatureManager>(sp => sp.GetRequiredService<FeatureManager>());
 
-            services.AddSingleton<ISessionManager, EmptySessionManager>();
+            services.AddScoped<FeatureManagerSnapshot>();
```

Comment:
What's the reasoning to go away from usage like `services.AddScoped<IFeatureManagerSnapshot, FeatureManagerSnapshot>();`

Reply:
I would expect `services.TryAddScoped<IFeatureManagerSnapshot, FeatureManagerSnapshot>()` should accomplish our needs.

Reply:
Same with registration of IFeatureManager.

Reply:
In the preview branch, we have IVariantFeatureManager and IFeatureManager sharing the same implementation FeatureManager, we use this pattern to register them. In the main branch, I agree to change it back to our previous way to register IFeatureManager.

Reply:
There is no overload of TryAddSingleton<TService, TImplementation>(ImplementationFactory). We need to use factory method to register FeatureManager implementation now. I think the current way to register IFeatureManager is better.